[GStreamer] Handle pre-roll in setVisibleInViewport()

ntrrgc opened the pull request:
#### aa09557dfddc2a32fe6c10f9c5e756b9cea48f61
<pre>
[GStreamer] Handle pre-roll in setVisibleInViewport()
<a href="https://bugs.webkit.org/show_bug.cgi?id=286750">https://bugs.webkit.org/show_bug.cgi?id=286750</a>

Reviewed by Philippe Normand.

Inside the player suspension code in setVisibleInViewport(), the
pipeline state is saved so that the pipeline can be set back to it when
later resuming.

However, the existing code was using the current state instead of the
target state. As a consequence, if the pipeline was still prerolling,
the pipeline would be set to READY instead of PAUSED on resume.

This patch fixes that issue by reading both the current and pending
state, and using the pending state if not VOID_PENDING.

Removing this unexpected transition to READY has reduced the amount of
crashes in presence of muted video and scrolling; however, I wouldn&apos;t go
so far as blaming the crashes on this bug, as changing the pipeline
state should not crash the pipeline.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::cancelLoad):
(WebCore::MediaPlayerPrivateGStreamer::loadNextLocation):
(WebCore::MediaPlayerPrivateGStreamer::setVisibleInViewport):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::setEosWithNoBuffers):

Canonical link: <a href="https://commits.webkit.org/289607@main">https://commits.webkit.org/289607@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2059a2960c2268d38289eb6a659afe76a1efce6b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87351 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/6864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/41721 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92212 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38089 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89402 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7224 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15007 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67496 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25227 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90353 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5517 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79052 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/47827 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5307 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33440 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37206 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75732 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34308 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94098 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14516 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10643 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76311 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/14755 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/74907 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75519 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18602 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/19887 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18311 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7453 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14534 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/19829 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14277 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/17721 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16060 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -3776,18 +3779,23 @@ void MediaPlayerPrivateGStreamer::setVisibleInViewport(bool isVisible)
         if (m_invisiblePlayerState != GST_STATE_VOID_PENDING) {
-            GST_DEBUG_OBJECT(pipeline(), "Element in viewport again, resuming playback.");
+            GST_DEBUG_OBJECT(pipeline(), "Element in viewport again, resuming playback via state change to %s.",
+                gst_element_state_get_name(m_invisiblePlayerState));
```
I read this block as "We're now visible again, so we restore the element state to whatever it had been previously requested but ignored, as it was visible".

Should `m_invisiblePlayerState` be set to `GST_STATE_VOID_PENDING` (or back to its default value)?

ntrrgc replied:
It shouldn't matter, but it could be added. Probably `m_isPausedByViewport` and `m_invisiblePlayerState` could be merged into one variable taking advantage of the `VOID_PENDING` pseudo-state.

I didn't want to make too many changes in the logic of this code to keep the change somewhat contained.

spenap commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
```diff
@@ -625,6 +625,7 @@ class MediaPlayerPrivateGStreamer
 
     bool m_didTryToRecoverPlayingState { false };
 
+    // The state the pipeline should be set back to after the player becomes visible in the viewport again.
```
Thanks for leaving this comment here.

It feels like the name `m_invisiblePlayerState` conveys what it does ("works as a temporary place to store the player state while invisible"), while your comment explains why it does it (So we can update the player state to a new one once visibility is restored).

A much longer but maybe suitable variable name could be `m_playerVisibilityRestoredState`?

ntrrgc replied:
> A much longer but maybe suitable variable name could be m_playerVisibilityRestoredState?

I'm not convinced... For once, it is not the "restored" state until it is actually restored. It would rather be `m_playerVisibilityStateToRestore` which is even longer.

It's hard to come up with a name, because the mechanism for pausing videos when scrolled away doesn't have an established name.

spenap replied:
Even if it is a long variable name, I feel there are benefits with going with `m_playerVisibilityStateToRestore`, tbh. I don't like the current name much...

ntrrgc replied:
How about `m_stateToRestoreWhenVisible`?

ntrrgc committed:
[GStreamer] Handle pre-roll in setVisibleInViewport()
https://bugs.webkit.org/show_bug.cgi?id=286750

Reviewed by Philippe Normand.

Inside the player suspension code in setVisibleInViewport(), the
pipeline state is saved so that the pipeline can be set back to it when
later resuming.

However, the existing code was using the current state instead of the
target state. As a consequence, if the pipeline was still prerolling,
the pipeline would be set to READY instead of PAUSED on resume.

This patch fixes that issue by reading both the current and pending
state, and using the pending state if not VOID_PENDING.

Removing this unexpected transition to READY has reduced the amount of
crashes in presence of muted video and scrolling; however, I wouldn't go
so far as blaming the crashes on this bug, as changing the pipeline
state should not crash the pipeline.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::cancelLoad):
(WebCore::MediaPlayerPrivateGStreamer::loadNextLocation):
(WebCore::MediaPlayerPrivateGStreamer::setVisibleInViewport):
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h:
* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::setEosWithNoBuffers):

Canonical link: https://commits.webkit.org/289607@main
